core/src/main/java/org/elasticsearch/index/engine/InternalEngine.java

```diff
@@ -174,15 +177,23 @@ public InternalEngine(EngineConfig engineConfig) throws EngineException {
                 switch (openMode) {
                     case OPEN_INDEX_AND_TRANSLOG:
                         writer = createWriter(false);
+                        String existingHistoryUUID = loadHistoryUUIDFromCommit(writer);
```

Comment:
Determining whether to generate a fresh or reuse the existing historyUUID should be done based on the ShardRouting.recoverySource() and can be set in the engine config:

EMPTY_STORE, SNAPSHOT, LOCAL_SHARDS => create fresh history ID
EXISTING_STORE, PEER => use existing history id (if available, otherwise create fresh)

Reply:
That's a great suggestion. I prefer to use it as a follow up PR when I actually fix the snapshot issue.

Reply:
I am ok with a followup. good catch @ywelsch